crates/sui-core/src/authority.rs

```diff
@@ -3738,6 +3740,96 @@ impl AuthorityState {
         (next_protocol_version, system_packages)
     }
 
+    /// Creates and executes an authenticator state tx, but does not commit the
+    /// results to the db.
+    ///
+    /// If the authenticator state object does not yet exist, it creates an
+    /// AuthenticatorStateCreate tx. Otherwise, it creates an AuthenticatorStateExpire tx.
+    ///
+    /// The effects of the tx are only written to the database after a certified
+    /// checkpoint has been formed and executed by CheckpointExecutor
+    pub async fn create_and_execute_authenticator_state_tx(
+        &self,
+        epoch_store: &Arc<AuthorityPerEpochStore>,
+        checkpoint: CheckpointSequenceNumber,
+    ) -> anyhow::Result<Option<TransactionEffects>> {
+        if !epoch_store.protocol_config().enable_jwk_consensus_updates() {
+            return Ok(None);
+        }
+
+        let authenticator_state_exists = epoch_store.authenticator_state_exists();
+        let tx = if authenticator_state_exists {
+            let next_epoch = epoch_store.epoch().checked_add(1).expect("epoch overflow");
+            let min_epoch =
+                next_epoch.saturating_sub(epoch_store.protocol_config().max_age_of_jwk_in_epochs());
+
+            let tx =
+                VerifiedTransaction::new_authenticator_state_expire(epoch_store.epoch(), min_epoch);
+
+            info!(
+                ?min_epoch,
+                tx_digest = ?tx.digest(),
+                "Creating AuthenticatorStateExpire tx",
+            );
+
+            tx
+        } else {
+            let tx = VerifiedTransaction::new_authenticator_state_create(epoch_store.epoch());
+            info!(tx_digest = ?tx.digest(), "Creating AuthenticatorStateCreate tx");
+            tx
+        };
+
+        let (effects, auth_state) = self
+            .execute_auxilliary_system_tx(
```

Comment:
It seems that if this function fails, we will fail checkpoint creation as well.

This could be fine, if the only source of transaction being executed could be state sync(e.g. it should be safe to abort checkpoint builder at this point).

But, I think this can happen also simply in the failover scenario - imagine this function executed and node crashed. At this point transaction is executed, checkpoint builder will re-enter this path and fail creating the checkpoint...


I am not sure if this is actually important (e.g. do we really care about crash resilience in once-a-lifetime scenario), but wanted to point out anyway

Reply:
This function never commits execution results to the DB. It only returns the effects so that we can construct the checkpoint. The commit happens only from checkpoint executor. So this should be safe. (And we also handle the case where the tx has already been executed via checkpoint at the point where we enter this function)